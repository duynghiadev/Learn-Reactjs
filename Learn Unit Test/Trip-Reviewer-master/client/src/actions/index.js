import axios from 'axios';

import {
  GET_TRIPS,
  GET_USERS,
  GET_TRIP_W_REVIEWER,
  CLEAR_TRIP_W_REVIEWER,
  GET_TRIP_W_REVIEWER_ERROR,
  USER_REGISTER,
  USER_LOGIN,
  RESET_PASSWORD,
  USER_AUTH,
  ADD_TRIP,
  UPDATE_TRIP,
  UPDATE_USER,
  DELETE_TRIP,
  GET_TRIP,
  GET_TRIP_ERROR,
  CLEAR_NEW_TRIP,
  CLEAR_TRIP,
  GET_USER_REVIEWS,
} from '../constants/action-names';

export function getTrips(limit = 10, start = 0, order = 'asc', list = []) {
  const request = axios.get(`/api/getManyTrips?limit=${limit}&skip=${start}&order=${order}`);

  return dispatch =>
    request.then(response => {
      dispatch({
        type: GET_TRIPS,
        payload: {
          list: [...list, ...response.data],
          newTripsCount: response.data.length,
        },
      });
    });
}

export function getTripWithReviewer(id) {
  const request = axios.get(`/api/getTrip?id=${id}`);

  return dispatch =>
    request
      .then(({ data: tripData }) => {
        const trip = tripData;

        return axios.get(`/api/getReviewer?id=${trip.ownerId}`).then(({ data: reviewerData }) => {
          const response = {
            trip,
            reviewer: reviewerData,
          };

          dispatch({
            type: GET_TRIP_W_REVIEWER,
            payload: response,
          });
        });
      })
      .catch(error =>
        dispatch({
          type: GET_TRIP_W_REVIEWER_ERROR,
          payload: error,
        })
      );
}

export function getUserReviews(userId) {
  const request = axios.get(`/api/getUserReviews?user=${userId}`);

  return dispatch =>
    request.then(response => {
      dispatch({
        type: GET_USER_REVIEWS,
        payload: response.data,
      });
    });
}

export function updateUser(data) {
  const request = axios.post('/api/userUpdate', data);

  return dispatch =>
    request.then(response => {
      dispatch({
        type: UPDATE_USER,
        payload: response.data,
      });
    });
}

export function getTrip(id) {
  const request = axios.get(`/api/getTrip?id=${id}`);

  return dispatch =>
    request
      .then(response => {
        dispatch({
          type: GET_TRIP,
          payload: response.data,
        });
      })
      .catch(error => {
        dispatch({
          type: GET_TRIP_ERROR,
          payload: error,
        });
      });
}

export function updateTrip(data) {
  const request = axios.post('/api/tripUpdate', data);

  return dispatch =>
    request.then(response => {
      dispatch({
        type: UPDATE_TRIP,
        payload: response.data,
      });
    });
}

export function deleteTrip(id) {
  const request = axios.delete(`/api/tripDelete?id=${id}`);

  return dispatch =>
    request.then(response => {
      dispatch({
        type: DELETE_TRIP,
        payload: response.data,
      });
    });
}

export function clearTrip() {
  return {
    type: CLEAR_TRIP,
    payload: {
      trip: null,
      updatedTrip: false,
      postDeleted: false,
    },
  };
}

export function clearTripWithReviewer() {
  return {
    type: CLEAR_TRIP_W_REVIEWER,
    payload: {
      trip: null,
      reviewer: null,
    },
  };
}

export function loginUser({ email, password }) {
  const request = axios.post('/api/login', { email, password });

  return dispatch =>
    request.then(response => {
      dispatch({
        type: USER_LOGIN,
        payload: response.data,
      });
    });
}

export function resetPassword({ email }) {
  const request = axios.post('/api/resetPassword', { email });

  return dispatch =>
    request.then(response => {
      dispatch({
        type: RESET_PASSWORD,
        payload: response.data,
      });
    });
}

export function auth() {
  const request = axios.get('/api/auth');

  return dispatch =>
    request.then(response => {
      dispatch({
        type: USER_AUTH,
        payload: response.data,
      });
    });
}

export function addTrip(trip) {
  const request = axios.post('/api/trip', trip);

  return dispatch =>
    request.then(response => {
      dispatch({
        type: ADD_TRIP,
        payload: response.data,
      });
    });
}

export function clearNewTrip() {
  return {
    type: CLEAR_NEW_TRIP,
    payload: null,
  };
}

export function getUsers() {
  const request = axios.get('/api/users');

  return dispatch =>
    request.then(response => {
      dispatch({
        type: GET_USERS,
        payload: response.data,
      });
    });
}

export function userRegister(user) {
  const request = axios.post('/api/register', user);

  return dispatch =>
    request.then(({ data }) => {
      dispatch({
        type: USER_REGISTER,
        payload: data,
      });
    });
}
