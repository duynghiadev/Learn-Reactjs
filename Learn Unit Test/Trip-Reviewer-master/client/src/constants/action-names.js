export const GET_TRIPS = 'get-trips';
export const GET_USERS = 'get-users';
export const ADD_TRIP = 'add-trip';
export const UPDATE_TRIP = 'update-trip';
export const UPDATE_USER = 'update-user';
export const DELETE_TRIP = 'delete-trip';
export const GET_TRIP = 'get-trip';
export const GET_TRIP_ERROR = 'get-trip-error';
export const GET_TRIP_W_REVIEWER = 'get-trip-w-reviewer';
export const GET_TRIP_W_REVIEWER_ERROR = 'get-trip-w-reviewer-error';
export const CLEAR_TRIP_W_REVIEWER = 'clear-trip-w-reviewer';
export const CLEAR_NEW_TRIP = 'clear-new-trip';
export const CLEAR_TRIP = 'clear-trip';
export const USER_REGISTER = 'user-register';
export const USER_LOGIN = 'user-login';
export const RESET_PASSWORD = 'reset-password';
export const USER_AUTH = 'user-auth';
export const GET_USER_REVIEWS = 'get-user-reviews';
