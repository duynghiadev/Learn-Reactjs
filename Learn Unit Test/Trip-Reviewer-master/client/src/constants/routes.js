export const HOME = '/';
export const LOGIN = '/login';
export const RESET = '/reset-password';
export const LOGOUT = '/user/logout';
export const REGISTER = '/user/register';
export const USER_PROFILE = '/user';
export const USER_REVIEWS = '/user/user-reviews';
export const EDIT_PROFILE = '/user/edit-profile';
export const EDIT_POST = '/user/edit-post';
export const ADD_TRIP = '/trip/add';
export const TRIPS = '/trips';
