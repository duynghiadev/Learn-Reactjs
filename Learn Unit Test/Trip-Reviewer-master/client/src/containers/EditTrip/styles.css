.edit-review-container {
  text-align: center;
}
.edit-review-container .form_element {
  margin-top: 20px;
}
.edit-review-container .title {
  font-size: 33px;
  font-weight: 300;
  color: #02bcd4;
}
.edit-review-container .form-input {
  font-size: 20px;
  border-radius: 3px;
  border: 1px solid #ebebeb;
  box-sizing: border-box;
  padding: 11px 31px;
  background: #f7f7f7;
  font-weight: 300;
  width: 80%;
}
.edit-review-container .label {
  display: block;
  text-align: left;
  margin-left: 10%;
  color: #02bcd4;
  font-size: 21px;
  margin-bottom: 3px;
}
.edit-review-container textarea {
  background: #f7f7f7;
  border: 1px solid #ebebeb;
  border-radius: 3px;
  width: 80%;
  padding: 20px;
  box-sizing: border-box;
  font-size: 14px;
}
.edit-review-container .add-button {
  cursor: pointer;
  font-size: 20px;
  padding: 10px 52px;
  background: #02bcd4;
  color: #fff;
  font-weight: 300;
  border-radius: 3px;
  border: none;
  margin-top: 50px;
  width: 80%;
}
.edit-review-container .add-button:hover {
  transform: scale(1.05);
  background: #009eb3;
}
.edit-review-container .add-button:active {
  transform: scale(1.01);
  background: #0495a8;
}
.edit-review-container .add-button:disabled {
  cursor: not-allowed;
  transform: none;
  background: grey;
}
.edit-review-container .error {
  color: #ff5722;
  margin-top: 5px;
}
.edit-review-container .delete-post {
  text-align: center;
  margin-top: 20px;
}
.edit-review-container .rdw-emoji-modal {
  left: -55px;
}
.edit-review-container .rdw-link-modal {
  left: -100px;
}
.edit-review-container .rdw-image-modal {
  left: -10px;
}
.edit-review-container .delete-post .button {
  cursor: pointer;
  font-size: 20px;
  padding: 10px 52px;
  background: #e91e63;
  color: #fff;
  font-weight: 300;
  border-radius: 3px;
  border: none;
  width: 80%;
  box-sizing: border-box;
  display: inline-block;
  margin-bottom: 50px;
}
.edit-review-container .delete-post .button:hover {
  transform: scale(1.05);
  background: #d11553;
}
.edit-review-container .delete-post .button:active {
  transform: scale(1.01);
  background: #ba124a;
}
.edit-review-container .delete-post .loading,
.edit-review-container .delete-post .loading:hover {
  cursor: not-allowed;
  transform: none;
  background: grey;
}
.edit-review-container .editor-wrapper {
  width: 80%;
  margin-left: 10%;
  border: 1px solid #ebebeb;
  background-color: #f7f7f7;
}
.edit-review-container .editor-self {
  padding-left: 10px;
  padding-right: 10px;
}
.edit-review-container .field-error {
  border: 2px solid #ff5722;
  color: #ff5722;
}
