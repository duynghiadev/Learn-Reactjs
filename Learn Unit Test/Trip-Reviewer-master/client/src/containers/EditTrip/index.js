import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Editor } from 'react-draft-wysiwyg';
import { EditorState, convertToRaw, convertFromRaw } from 'draft-js';

import StarsRating from '../../components/StarsRating';
import CountrySelector from '../../components/CountrySelector';
import { getTrip, updateTrip, clearTrip, deleteTrip } from '../../actions';
import * as routes from '../../constants/routes';
import toolbar from '../../constants/toolbar';

import './styles.css';

class EditTrip extends PureComponent {
  state = {
    editorState: EditorState.createEmpty(),
    formdata: {
      _id: this.props.match.params.id,
      title: '',
      country: '',
      description: '',
      duration: '',
      rating: 0,
      expences: '',
    },
    hideError: {
      title: false,
      country: false,
      description: false,
      duration: false,
      rating: false,
      expences: false,
    },
    loading: false,
  };

  componentWillMount() {
    this.props.dispatch(getTrip(this.props.match.params.id));
  }

  componentWillReceiveProps(nextProps) {
    const { trip, updatedTrip } = nextProps.trips;

    if (trip && trip.error) {
      nextProps.history.push('/Not-Found');
    } else if (updatedTrip && updatedTrip.success) {
      nextProps.history.push(`${routes.TRIPS}/${trip._id}`);
    } else if (updatedTrip) {
      this.setState({
        loading: false,
        hideError: {
          title: false,
          country: false,
          description: false,
          duration: false,
          rating: false,
          expences: false,
        },
      });
    } else if (trip) {
      let editorState = EditorState.createEmpty();
      try {
        const parsedDescription = JSON.parse(trip.description);
        const contantState = convertFromRaw(parsedDescription);
        editorState = EditorState.createWithContent(contantState);
      } catch (e) {
        console.log('Parsing description error: ', e); // eslint-disable-line no-console
        console.log('Object: ', trip.description); // eslint-disable-line no-console
      }

      this.setState({
        editorState,
        loading: false,
        formdata: {
          _id: trip._id,
          title: trip.title,
          country: trip.country,
          description: trip.description,
          duration: trip.duration,
          rating: trip.rating,
          expences: trip.expences,
        },
      });
    }
  }

  componentWillUnmount() {
    this.props.dispatch(clearTrip());
  }

  onEditorStateChange = editorState => {
    const newFormdata = { ...this.state.formdata };
    const newHideError = {
      ...this.state.hideError,
      description: true,
    };
    const contentState = editorState.getCurrentContent();
    const rawState = convertToRaw(contentState);

    newFormdata.description = JSON.stringify(rawState);

    this.setState({
      editorState,
      formdata: newFormdata,
      hideError: newHideError,
    });
  };

  getErrorClass(fieldName) {
    return this.formFieldHasError(fieldName) ? 'field-error' : '';
  }

  formFieldHasError(fieldName) {
    const { updatedTrip } = this.props.trips;
    const { hideError } = this.state;
    const errors = updatedTrip && updatedTrip.error && updatedTrip.error.errors;

    return errors && errors[fieldName] && !hideError[fieldName];
  }

  handleInput = event => {
    const newFormdata = { ...this.state.formdata };
    const newHideError = { ...this.state.hideError };
    const { value, name } = event.target;

    newFormdata[name] = value;
    newHideError[name] = true;

    this.setState({
      formdata: newFormdata,
      hideError: newHideError,
    });
  };

  handleCountrychange = countryObject => {
    const newFormdata = {
      ...this.state.formdata,
      country: countryObject.countryName,
    };

    this.setState({
      formdata: newFormdata,
    });
  };

  handleRating = rating => {
    const newFormdata = {
      ...this.state.formdata,
      rating,
    };
    const newHideError = {
      ...this.state.hideError,
      rating: true,
    };

    this.setState({
      formdata: newFormdata,
      hideError: newHideError,
    });
  };

  submitForm = event => {
    event.preventDefault();
    this.props.dispatch(updateTrip(this.state.formdata));
    this.setState({ loading: true });
  };

  deletePost = () => {
    if (!this.state.loading) {
      this.props.dispatch(deleteTrip(this.props.match.params.id));
      this.setState({ loading: true });
    }
  };

  goToReviews = () => {
    this.props.history.push(routes.USER_REVIEWS);
  };

  redirectUser = () => {
    setTimeout(this.goToReviews, 1000);
  };

  renderError(fieldName) {
    if (this.formFieldHasError(fieldName)) {
      const error = this.props.trips.updatedTrip.error.errors[fieldName];

      return <div className="error">{error.message}</div>;
    }
    return null;
  }

  renderPostDeleted() {
    if (this.props.trips.postDeleted) {
      return (
        <div className="red_tag">
          Post Deleted
          {this.redirectUser()}
        </div>
      );
    }
    return null;
  }

  render() {
    const { loading, formdata } = this.state;
    const { title, country, duration, rating, expences } = formdata;

    return (
      <div className="edit-review-container limited-width">
        {this.renderPostDeleted()}

        <form onSubmit={this.submitForm}>
          <h2 className="title">Edit story</h2>

          <div className="form_element">
            <span className="label">Title:</span>
            <input
              type="text"
              name="title"
              className={`form-input ${this.getErrorClass('title')}`}
              placeholder="Enter title"
              value={title}
              onChange={this.handleInput}
            />
          </div>
          {this.renderError('title')}

          <div className="form_element">
            <span className="label">Trip to:</span>
            <CountrySelector
              defaultCountry={country}
              getSelectedCountry={this.handleCountrychange}
            />
          </div>

          <div className="form_element">
            <span className="label">Description:</span>
            <Editor
              editorState={this.state.editorState}
              wrapperClassName={`editor-wrapper ${this.getErrorClass('description')}`}
              editorClassName="editor-self"
              onEditorStateChange={this.onEditorStateChange}
              toolbar={toolbar}
            />
          </div>
          {this.renderError('description')}

          <div className="form_element">
            <span className="label">Duration (days):</span>
            <input
              type="number"
              name="duration"
              className={`form-input ${this.getErrorClass('duration')}`}
              placeholder="Enter duration"
              value={duration}
              onChange={this.handleInput}
            />
          </div>
          {this.renderError('duration')}

          <div className="form_element">
            <span className="label">Expenses ($):</span>
            <input
              type="number"
              name="expences"
              className={`form-input ${this.getErrorClass('expences')}`}
              placeholder="Enter expenses"
              value={expences}
              onChange={this.handleInput}
            />
          </div>
          {this.renderError('expences')}

          <div className="form_element">
            <StarsRating
              rating={rating}
              error={this.getErrorClass('rating')}
              onChange={this.handleRating}
            />
          </div>
          {this.renderError('rating')}

          <button type="submit" className="add-button" disabled={loading}>
            Edit story
          </button>
          <div className="delete-post">
            <div className={`button${loading ? ' loading' : ''}`} onClick={this.deletePost}>
              Delete story
            </div>
          </div>
        </form>
      </div>
    );
  }
}

EditTrip.propTypes = {
  dispatch: PropTypes.func.isRequired,
  match: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  trips: PropTypes.object.isRequired,
};

function mapStateToProps(state) {
  return {
    trips: state.trips,
  };
}

export default connect(mapStateToProps)(EditTrip);
