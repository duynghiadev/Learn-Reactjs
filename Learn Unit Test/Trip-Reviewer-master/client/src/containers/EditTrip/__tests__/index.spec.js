import React from 'react';
import configureStore from 'redux-mock-store';
import { create } from 'react-test-renderer';
import { shallow } from 'enzyme';

import EditTrip from '..';
import { GET_TRIP } from '../../../constants/action-names';
import { updateTrip, clearTrip, deleteTrip } from '../../../actions';
import * as routes from '../../../constants/routes';

jest.mock('react-router-dom', () => ({ Link: 'Link' }));
jest.mock('react-draft-wysiwyg', () => ({ Editor: 'Editor' }));
jest.mock('draft-js', () => ({
  EditorState: {
    createEmpty: () => ({}),
    createWithContent: () => ({}),
  },
  convertFromRaw: () => ({}),
  convertToRaw: () => ({}),
}));
jest.mock('../../../constants/toolbar', () => ({}));
jest.mock('./../../../components/StarsRating', () => 'StarsRating');
jest.mock('./../../../components/CountrySelector', () => 'CountrySelector');
jest.mock('./../../../actions', () => ({
  getTrip: jest.fn(() => ({
    type: GET_TRIP,
    payload: {},
  })),
  updateTrip: jest.fn(() => ({
    type: 'updateTrip',
  })),
  clearTrip: jest.fn(() => ({
    type: 'clearTrip',
  })),
  deleteTrip: jest.fn(() => ({
    type: 'deleteTrip',
  })),
}));

const mockStore = configureStore();

const mockComponent = (initialState = {}, props) => {
  const store = mockStore(initialState);

  return <EditTrip {...props} store={store} />;
};

describe('<EditTrip />', () => {
  it('should render component', () => {
    const initialState = {
      trips: {},
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
    };
    const tree = create(mockComponent(initialState, props)).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render component with updatedTrip and errors', () => {
    const initialState = {
      trips: {
        updatedTrip: {
          success: false,
          error: {
            errors: {
              title: { message: 'title error' },
              description: { message: 'description error' },
              duration: { message: 'duration error' },
              rating: { message: 'rating error' },
              expences: { message: 'expences error' },
            },
          },
        },
        trip: {
          _id: '_id',
        },
      },
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
    };
    const tree = create(mockComponent(initialState, props)).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render component with postDeleted', () => {
    const initialState = {
      trips: {
        postDeleted: true,
      },
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
    };
    const tree = create(mockComponent(initialState, props)).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should change state when componentWillReceiveProps is called with a correct description', () => {
    const initialState = {
      trips: {},
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
    };
    const nextProps = {
      trips: {
        trip: {
          _id: 'trip._id',
          title: 'trip.title',
          country: 'trip.country',
          description: '{"test":"test"}',
          duration: 'trip.duration',
          rating: 'trip.rating',
          expences: 'trip.expences',
        },
      },
    };

    const instance = shallow(mockComponent(initialState, props)).dive().instance();

    instance.componentWillReceiveProps(nextProps);

    expect(instance.state.formdata).toEqual(nextProps.trips.trip);
  });

  it('should change state when componentWillReceiveProps is called with a wrong description', () => {
    const initialState = {
      trips: {},
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
    };
    const nextProps = {
      trips: {
        trip: {
          _id: 'trip._id',
          title: 'trip.title',
          country: 'trip.country',
          description: 'trip.description',
          duration: 'trip.duration',
          rating: 'trip.rating',
          expences: 'trip.expences',
        },
      },
    };

    const instance = shallow(mockComponent(initialState, props)).dive().instance();

    instance.componentWillReceiveProps(nextProps);

    expect(instance.state.formdata).toEqual(nextProps.trips.trip);
  });

  it('should not change state when componentWillReceiveProps is called with no trip', () => {
    const initialState = {
      trips: {},
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
    };
    const nextProps = {
      trips: {},
    };

    const instance = shallow(mockComponent(initialState, props)).dive().instance();

    instance.componentWillReceiveProps(nextProps);

    expect(instance.state.formdata).not.toEqual(nextProps.trips.trip);
  });

  it('should redirect when componentWillReceiveProps is called with error', () => {
    const push = jest.fn();
    const path = '/Not-Found';
    const initialState = {
      trips: {},
    };
    const props = {
      match: {
        params: { id: 'id' },
      },
    };
    const nextProps = {
      trips: {
        trip: { error: 'error' },
      },
      history: { push },
    };

    const instance = shallow(mockComponent(initialState, props)).dive().instance();

    instance.componentWillReceiveProps(nextProps);

    expect(push).toHaveBeenCalledWith(path);
  });

  it('should redirect when componentWillReceiveProps is called with updatedTrip success', () => {
    const push = jest.fn();
    const tripId = 'test';
    const path = `${routes.TRIPS}/${tripId}`;
    const initialState = {
      trips: {},
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
    };
    const nextProps = {
      trips: {
        trip: {
          _id: tripId,
        },
        updatedTrip: {
          success: true,
        },
      },
      history: {
        push,
      },
    };

    const instance = shallow(mockComponent(initialState, props)).dive().instance();

    instance.componentWillReceiveProps(nextProps);

    expect(push).toHaveBeenCalledWith(path);
  });

  it('should not redirect when componentWillReceiveProps is called with updatedTrip success false', () => {
    const push = jest.fn();
    const setState = jest.fn();
    const tripId = 'test';
    const path = `${routes.TRIPS}/${tripId}`;
    const initialState = {
      trips: {},
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
    };
    const nextProps = {
      trips: {
        trip: {
          _id: tripId,
        },
        updatedTrip: {
          success: false,
        },
      },
      history: {
        push,
      },
    };
    const expected = {
      loading: false,
      hideError: {
        title: false,
        country: false,
        description: false,
        duration: false,
        rating: false,
        expences: false,
      },
    };

    const instance = shallow(mockComponent(initialState, props)).dive().instance();
    instance.setState = setState;

    instance.componentWillReceiveProps(nextProps);

    expect(push).not.toHaveBeenCalledWith(path);
    expect(setState).toHaveBeenCalledWith(expected);
  });

  it('should dispatch clearTrip when componentWillUnmount is called', () => {
    const initialState = {
      trips: {},
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
    };

    const instance = shallow(mockComponent(initialState, props)).dive().instance();

    instance.componentWillUnmount();

    expect(clearTrip).toHaveBeenCalled();
  });

  it('should dispatch updateTrip when submitForm is called', () => {
    const preventDefault = jest.fn();
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
    };
    const state = {
      country: 'country',
      duration: 'duration',
      ownerId: 'id',
      expences: 'expences',
      rating: 'rating',
      description: 'description',
      title: 'title',
    };
    const initialState = {
      trips: {},
    };
    const event = { preventDefault };

    const instance = shallow(mockComponent(initialState, props)).dive().instance();
    instance.setState({ formdata: state });

    instance.submitForm(event);

    expect(preventDefault).toHaveBeenCalled();
    expect(updateTrip).toHaveBeenCalledWith(state);
  });

  it('should dispatch deleteTrip when deletePost is called', () => {
    const setState = jest.fn();
    const initialState = {
      trips: {},
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
    };

    const instance = shallow(mockComponent(initialState, props)).dive().instance();
    instance.setState = setState;

    instance.deletePost();

    expect(deleteTrip).toHaveBeenCalledWith('id');
    expect(setState).toHaveBeenCalledWith({ loading: true });
  });

  it('should not dispatch deleteTrip when deletePost is called while loading', () => {
    const setState = jest.fn();
    const dispatch = jest.fn();
    const initialState = {
      trips: {},
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
      dispatch,
    };

    const instance = shallow(mockComponent(initialState, props)).dive().instance();
    instance.setState = setState;
    instance.state.loading = true;

    instance.deletePost();

    expect(dispatch).not.toHaveBeenCalled();
    expect(setState).not.toHaveBeenCalledWith({ loading: true });
  });

  it('handleInput should change state correctly', () => {
    const initialState = {
      trips: {},
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
    };
    const event = {
      target: {
        value: 'test',
        name: 'title',
      },
    };

    const instance = shallow(mockComponent(initialState, props)).dive().instance();

    instance.handleInput(event);

    expect(instance.state.formdata.title).toEqual('test');
  });

  it('onEditorStateChange should change state correctly', () => {
    const initialState = {
      trips: {},
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
    };
    const editorState = {
      getCurrentContent: () => ({}),
      test: 'test',
    };

    const instance = shallow(mockComponent(initialState, props)).dive().instance();

    instance.onEditorStateChange(editorState);

    expect(instance.state.editorState).toEqual(editorState);
    expect(instance.state.formdata.description).toEqual('{}');
  });

  it('handleRating should change state correctly', () => {
    const initialState = {
      trips: {},
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
    };
    const rating = 3;

    const instance = shallow(mockComponent(initialState, props)).dive().instance();

    instance.handleRating(rating);

    expect(instance.state.formdata.rating).toEqual(rating);
  });

  it('handleCountrychange should change state correctly', () => {
    const initialState = {
      trips: {},
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
    };
    const countryName = 'countryName';
    const coutryObject = { countryName };

    const instance = shallow(mockComponent(initialState, props)).dive().instance();

    instance.handleCountrychange(coutryObject);

    expect(instance.state.formdata.country).toEqual(countryName);
  });

  it('goToReviews should redirect correctly', () => {
    const push = jest.fn();
    const path = routes.USER_REVIEWS;
    const initialState = {
      trips: {},
    };
    const props = {
      match: {
        params: {
          id: 'id',
        },
      },
      history: {
        push,
      },
    };

    const instance = shallow(mockComponent(initialState, props)).dive().instance();

    instance.goToReviews();

    expect(push).toHaveBeenCalledWith(path);
  });
});
