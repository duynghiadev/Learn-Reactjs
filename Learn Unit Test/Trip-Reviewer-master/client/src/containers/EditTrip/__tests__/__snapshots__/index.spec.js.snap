// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditTrip /> should render component 1`] = `
<div
  className="edit-review-container limited-width"
>
  <form
    onSubmit={[Function]}
  >
    <h2
      className="title"
    >
      Edit story
    </h2>
    <div
      className="form_element"
    >
      <span
        className="label"
      >
        Title:
      </span>
      <input
        className="form-input "
        name="title"
        onChange={[Function]}
        placeholder="Enter title"
        type="text"
        value=""
      />
    </div>
    <div
      className="form_element"
    >
      <span
        className="label"
      >
        Trip to:
      </span>
      <CountrySelector
        defaultCountry=""
        getSelectedCountry={[Function]}
      />
    </div>
    <div
      className="form_element"
    >
      <span
        className="label"
      >
        Description:
      </span>
      <Editor
        editorClassName="editor-self"
        editorState={Object {}}
        onEditorStateChange={[Function]}
        toolbar={Object {}}
        wrapperClassName="editor-wrapper "
      />
    </div>
    <div
      className="form_element"
    >
      <span
        className="label"
      >
        Duration (days):
      </span>
      <input
        className="form-input "
        name="duration"
        onChange={[Function]}
        placeholder="Enter duration"
        type="number"
        value=""
      />
    </div>
    <div
      className="form_element"
    >
      <span
        className="label"
      >
        Expenses ($):
      </span>
      <input
        className="form-input "
        name="expences"
        onChange={[Function]}
        placeholder="Enter expenses"
        type="number"
        value=""
      />
    </div>
    <div
      className="form_element"
    >
      <StarsRating
        error=""
        onChange={[Function]}
        rating={0}
      />
    </div>
    <button
      className="add-button"
      disabled={false}
      type="submit"
    >
      Edit story
    </button>
    <div
      className="delete-post"
    >
      <div
        className="button"
        onClick={[Function]}
      >
        Delete story
      </div>
    </div>
  </form>
</div>
`;

exports[`<EditTrip /> should render component with postDeleted 1`] = `
<div
  className="edit-review-container limited-width"
>
  <div
    className="red_tag"
  >
    Post Deleted
  </div>
  <form
    onSubmit={[Function]}
  >
    <h2
      className="title"
    >
      Edit story
    </h2>
    <div
      className="form_element"
    >
      <span
        className="label"
      >
        Title:
      </span>
      <input
        className="form-input "
        name="title"
        onChange={[Function]}
        placeholder="Enter title"
        type="text"
        value=""
      />
    </div>
    <div
      className="form_element"
    >
      <span
        className="label"
      >
        Trip to:
      </span>
      <CountrySelector
        defaultCountry=""
        getSelectedCountry={[Function]}
      />
    </div>
    <div
      className="form_element"
    >
      <span
        className="label"
      >
        Description:
      </span>
      <Editor
        editorClassName="editor-self"
        editorState={Object {}}
        onEditorStateChange={[Function]}
        toolbar={Object {}}
        wrapperClassName="editor-wrapper "
      />
    </div>
    <div
      className="form_element"
    >
      <span
        className="label"
      >
        Duration (days):
      </span>
      <input
        className="form-input "
        name="duration"
        onChange={[Function]}
        placeholder="Enter duration"
        type="number"
        value=""
      />
    </div>
    <div
      className="form_element"
    >
      <span
        className="label"
      >
        Expenses ($):
      </span>
      <input
        className="form-input "
        name="expences"
        onChange={[Function]}
        placeholder="Enter expenses"
        type="number"
        value=""
      />
    </div>
    <div
      className="form_element"
    >
      <StarsRating
        error=""
        onChange={[Function]}
        rating={0}
      />
    </div>
    <button
      className="add-button"
      disabled={false}
      type="submit"
    >
      Edit story
    </button>
    <div
      className="delete-post"
    >
      <div
        className="button"
        onClick={[Function]}
      >
        Delete story
      </div>
    </div>
  </form>
</div>
`;

exports[`<EditTrip /> should render component with updatedTrip and errors 1`] = `
<div
  className="edit-review-container limited-width"
>
  <form
    onSubmit={[Function]}
  >
    <h2
      className="title"
    >
      Edit story
    </h2>
    <div
      className="form_element"
    >
      <span
        className="label"
      >
        Title:
      </span>
      <input
        className="form-input field-error"
        name="title"
        onChange={[Function]}
        placeholder="Enter title"
        type="text"
        value=""
      />
    </div>
    <div
      className="error"
    >
      title error
    </div>
    <div
      className="form_element"
    >
      <span
        className="label"
      >
        Trip to:
      </span>
      <CountrySelector
        defaultCountry=""
        getSelectedCountry={[Function]}
      />
    </div>
    <div
      className="form_element"
    >
      <span
        className="label"
      >
        Description:
      </span>
      <Editor
        editorClassName="editor-self"
        editorState={Object {}}
        onEditorStateChange={[Function]}
        toolbar={Object {}}
        wrapperClassName="editor-wrapper field-error"
      />
    </div>
    <div
      className="error"
    >
      description error
    </div>
    <div
      className="form_element"
    >
      <span
        className="label"
      >
        Duration (days):
      </span>
      <input
        className="form-input field-error"
        name="duration"
        onChange={[Function]}
        placeholder="Enter duration"
        type="number"
        value=""
      />
    </div>
    <div
      className="error"
    >
      duration error
    </div>
    <div
      className="form_element"
    >
      <span
        className="label"
      >
        Expenses ($):
      </span>
      <input
        className="form-input field-error"
        name="expences"
        onChange={[Function]}
        placeholder="Enter expenses"
        type="number"
        value=""
      />
    </div>
    <div
      className="error"
    >
      expences error
    </div>
    <div
      className="form_element"
    >
      <StarsRating
        error="field-error"
        onChange={[Function]}
        rating={0}
      />
    </div>
    <div
      className="error"
    >
      rating error
    </div>
    <button
      className="add-button"
      disabled={false}
      type="submit"
    >
      Edit story
    </button>
    <div
      className="delete-post"
    >
      <div
        className="button"
        onClick={[Function]}
      >
        Delete story
      </div>
    </div>
  </form>
</div>
`;
